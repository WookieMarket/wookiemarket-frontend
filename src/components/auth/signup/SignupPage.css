.container-form {
  width: 30%;
  height: 45vh;
  /* Establece la altura del contenedor en el 50% de la altura de la ventana */
  position: absolute;
  /* Cambia a posici√≥n absoluta */
  top: 55%;
  /* Alinea el borde superior del contenedor al 50% de la ventana */
  left: 50%;
  /* Alinea el borde izquierdo del contenedor al 50% de la ventana */
  transform: translate(-50%, -50%);
  /* Centra vertical y horizontalmente */
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  padding: 30px 10px 20px 10px;
  gap: 10px;
  border-radius: 10px;
  text-align: center;
  border: 2px solid var(--border-color-form);
  background-color: var(--background-color-form);
}

.container-form .form-group {
  width: 100%;
  padding: 0 15px;
  display: flex;
  flex-direction: column;
  gap: 10px;
}

.container-form .email-label {
  /* font-family: var(--font-family); */
  font-size: 13px;
  margin-bottom: 10px;
}

.container-form .password-label {
  /* font-family: var(--font-family); */
  font-size: 13px;
  margin-bottom: 10px;
}

.container-form .password-input {
  height: 20px;
  width: 100%;
  /* font-family: var(--font-family); */
  margin: auto;
  font-size: 12px;
  margin-bottom: 10px;
  border-radius: 6px;
  border: 3px solid gainsboro;
}

.container-form .email-input {
  height: 20px;
  width: 100%;
  margin: auto;
  font-size: 13px;
  margin-bottom: 10px;
  border-radius: 6px;
  border: 3px solid gainsboro;
}