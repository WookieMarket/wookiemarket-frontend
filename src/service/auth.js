import client, {
  removeAuthorizationHeader,
  setAuthorizationHeader,
} from '../api/client';
import storage from '../utils/storage';

export const signup = userData => {
  return client.post('/api/auth/signup', userData).then(({ jwt }) => {
    setAuthorizationHeader(jwt);
    storage.set('auth', jwt);
  });
};

export const login = credentials => {
  return client.post('/api/auth/login', credentials).then(({ jwt }) => {
    setAuthorizationHeader(jwt);
    if (credentials.rememberMe) {
      storage.set('auth', jwt);
    }
    console.log('jwt,', jwt);
    return jwt;
  });
};

export const logout = () => {
  return Promise.resolve().then(() => {
    removeAuthorizationHeader();
    storage.remove('auth');
  });
};

export const emailPassword = email => {
  return client.post('api/users/email-password', { to: email });
};

export const resetPassword = (email, newPassword, token) => {
  return client.post('api/users/recover-password', {
    email,
    newPassword,
    token,
  });
};

export const deleteUser = email => {
  return client.post('api/users/deleted-user', { email: email });
};
