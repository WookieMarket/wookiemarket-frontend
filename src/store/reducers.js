export {default as adverts} from './slices/adverts';
export {default as ui} from './slices/ui';