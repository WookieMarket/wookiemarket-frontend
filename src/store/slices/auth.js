import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

// use rejectWithValue to take care of the error

export const authSignup = createAsyncThunk(
  'auth/signup',
  async (userData, { extra: { service }, rejectWithValue }) => {
    try {
      await service.auth.signup(userData);
    } catch (error) {
      return rejectWithValue(error);
    }
  },
);

export const authLogin = createAsyncThunk(
  'auth/login',
  async (credentials, { extra: { service }, rejectWithValue }) => {
    try {
      const response = await service.auth.login(credentials);
      return response;
    } catch (error) {
      return rejectWithValue(error);
    }
  },
);

export const authLogout = createAsyncThunk(
  'auth/logout',
  (_, { extra: { service } }) => service.auth.logout(),
);

export const emailResetPassword = createAsyncThunk(
  'auth/emailResetPassword',
  async (email, { extra: { service }, rejectWithValue }) => {
    try {
      const resetEmail = await service.auth.emailPassword(email);
      return resetEmail;
    } catch (error) {
      return rejectWithValue(error);
    }
  },
);

export const resetPassword = createAsyncThunk(
  'auth/resetPassword',
  async (
    { email, newPassword, token },
    { extra: { service }, rejectWithValue },
  ) => {
    try {
      const response = await service.auth.resetPassword(
        email,
        newPassword,
        token,
      );
      return response;
    } catch (error) {
      return rejectWithValue(error);
    }
  },
);

export const deleteAccount = createAsyncThunk(
  'auth/deleteAccount',
  async (email, { extra: { service }, rejectWithValue }) => {
    try {
      const response = await service.auth.deleteUser(email);
      return response;
    } catch (error) {
      return rejectWithValue(error);
    }
  },
);

const auth = createSlice({
  name: 'auth',
  initialState: {
    auth: false,
    jwt: '',
  },
  extraReducers: builder => {
    builder
      .addCase(authSignup.fulfilled, (state, action) => {
        state.auth = true;
        state.jwt = '';
      })
      .addCase(authLogin.fulfilled, (state, action) => {
        state.auth = true;
        state.jwt = action.payload;
      })
      .addCase(authLogout.fulfilled, (state, action) => {
        state.auth = false;
        state.jwt = '';
      })
      .addCase(deleteAccount.fulfilled, (state, action) => {
        state.auth = false;
        state.jwt = '';
      });
  },
});

export default auth.reducer;
