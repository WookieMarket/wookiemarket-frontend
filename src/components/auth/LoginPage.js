import { useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { getUi } from "../../store/selectors";
import { resetError } from "../../store/slices/ui";
import { authLogin } from "../../store/slices/auth";
import ErrorModal from "../shared/modal/ErrorModal";
import Spinner from "../shared/spinner/Spinner";
import Layout from "../layout/Layout";

//DONE Log in with username and password and a checkbox to give the option to persist the token, also handle errors and user feedback. When doing Login I want to send the user to the page they wanted to go to.

function LoginPage() {
  const dispatch = useDispatch();
  const { isLoading, error } = useSelector(getUi);

  const [credentials, setCredentials] = useState({
    username: "",
    password: "",
    rememberMe: false,
  });

  const handleErrorClick = () => {
    dispatch(resetError());
  };

  const handleChange = event => {
    const { name, value, type, checked } = event.target;

    setCredentials(prevCredentials => ({
      ...prevCredentials,
      [name]: type === "checkbox" ? checked : value,
    }));
  };

  const handleSubmit = async event => {
    event.preventDefault();

    dispatch(authLogin(credentials));
  };

  const buttonDisabled =
    isLoading || !credentials.username || !credentials.password;

  return (
    <Layout title="Login Page">
      <div>
        {isLoading ? (
          <Spinner message="cargando..." />
        ) : (
          <form onSubmit={handleSubmit} className="container-form">
            <label className="form-label" htmlFor="username">
              Username
            </label>
            <input
              id="username"
              className="form-input"
              type="text"
              name="username"
              onChange={handleChange}
              value={credentials.username}
              required
            />
            <label className="form-label" htmlFor="password">
              Password
            </label>
            <input
              id="password"
              className="form-input"
              type="password"
              name="password"
              onChange={handleChange}
              value={credentials.password}
              required
            />
            <label htmlFor="rememberMe">
              rememberMe
              <input
                id="rememberMe"
                type="checkbox"
                name="rememberMe"
                checked={credentials.rememberMe}
                onChange={handleChange}
              />
            </label>
            <button
              data-testid="button"
              type="submit"
              //variant="primary"
              width="button-form"
              disabled={buttonDisabled}>
              Log in
            </button>
          </form>
        )}

        {error && (
          <ErrorModal
            title="Error"
            message={error.message}
            onCancel={handleErrorClick}
          />
        )}
      </div>
    </Layout>
  );
}

export default LoginPage;
