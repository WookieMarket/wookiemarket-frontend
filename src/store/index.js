import { configureStore as rtkConfigureStore } from '@reduxjs/toolkit';
import * as service from '../service';
import * as reducers from './reducers';
import { failureRedirects, successRedirects } from './middleware';
import { authSignup, authLogin } from './slices/auth';

export default function configureStore(preloadedState, { router }) {
  const extraMiddleware = [
    failureRedirects(router, { 401: '/login', 404: '/404' }),
    successRedirects(router, {
      [authSignup.fulfilled.type]: () => {
        return router.state.location.state?.from?.pathname || '/';
      },
      [authLogin.fulfilled.type]: () => {
        return router.state.location.state?.from?.pathname || '/';
      },
    }),
  ];

  const store = rtkConfigureStore({
    reducer: reducers,
    middleware: getDefaultMiddleware =>
      getDefaultMiddleware({
        thunk: { extraArgument: { service, router } },
        serializableCheck: false,
      }).concat(extraMiddleware),
    preloadedState,
  });
  return store;
}
