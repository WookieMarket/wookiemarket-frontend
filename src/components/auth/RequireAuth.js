import { useSelector } from 'react-redux';
import { getIsLogged } from '../../store/selectors';
import { Navigate, useLocation } from 'react-router-dom';

//NOTE The children property refers to the element that is inside the tag

//DONE Function that I use to cover a component if it is required to have authorization, I pass a location object through the state property to know which route I come from

const RequireAuth = ({ children }) => {
  const isLogged = useSelector(getIsLogged);
  const location = useLocation();

  if (!isLogged) {
    return <Navigate to="/login" state={{ from: location }} />;
  }
  return children;
};

export default RequireAuth;
