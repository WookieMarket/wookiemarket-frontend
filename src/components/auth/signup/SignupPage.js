import { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { getUi } from '../../../store/selectors';
import { resetError } from '../../../store/slices/ui';
import { authSignup } from '../../../store/slices/auth';
import ErrorModal from '../../shared/modal/ErrorModal';
import Spinner from '../../shared/spinner/Spinner';
import Layout from '../../layout/Layout';
import { useTranslation } from 'react-i18next';
import Form from '../../shared/form/Form';
import './SignupPage.css';

/**
 * This page allows user to create an account.
 * If account is successfully created a session is created
 * and user is forwarded to home page.
 */
function SignupPage() {
  const { t } = useTranslation();
  const dispatch = useDispatch();
  const { isLoading, error } = useSelector(getUi);

  const [userData, setUserData] = useState({
    email: '',
    username: '',
    password: '',
  });

  const handleErrorClick = () => {
    dispatch(resetError());
  };

  const handleChange = event => {
    setUserData(prevUserData => ({
      ...prevUserData,
      [event.target.name]: event.target.value,
    }));
  };

  const handleSubmit = async event => {
    event.preventDefault();

    dispatch(authSignup(userData));
  };

  const buttonDisabled =
    isLoading || !userData.username || !userData.password || !userData.email;

  return (
    <Layout title={t('Signup Page')}>
      <div>
        {isLoading ? (
          <Spinner message={t('charging...')} />
        ) : (
          <form onSubmit={handleSubmit} className="container-form">
            <Form
              classNameLabel="email-label"
              htmlFor="email"
              text={t('Email')}
              classNameInput="email-input"
              inputId="email"
              inputType="text"
              inputName="email"
              value={userData.email}
              handleChange={handleChange}
              placeholder={t('Email')}
              required
            />
            <Form
              classNameLabel="user-label"
              htmlFor="username"
              text={t('Username')}
              classNameInput="user-input"
              inputId="username"
              inputType="text"
              inputName="username"
              value={userData.username}
              handleChange={handleChange}
              placeholder={t('Username')}
              required
            />

            <Form
              classNameLabel="password-label"
              htmlFor="password"
              text={t('Password')}
              classNameInput="password-input"
              inputId="password"
              inputType="password"
              inputName="password"
              value={userData.password}
              handleChange={handleChange}
              placeholder={t('Password')}
              required
            />
            <button
              data-testid="button"
              type="submit"
              width="button-form"
              disabled={buttonDisabled}
            >
              {t('Register-button-title')}
            </button>
          </form>
        )}

        {error && (
          <ErrorModal
            title="Error"
            message={error.data.error}
            onCancel={handleErrorClick}
            testid="modalSignup"
          />
        )}
      </div>
    </Layout>
  );
}

export default SignupPage;
