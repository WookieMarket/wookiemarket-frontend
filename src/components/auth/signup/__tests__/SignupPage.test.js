import { render, screen } from '@testing-library/react';
import SignupPage from '../SignupPage';
import userEvent from '@testing-library/user-event';
import { authSignup } from '../../../../store/slices/auth';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import * as defaultState from '../../../../store/reducers';

jest.mock('../../../../store/slices/auth', () => ({
  __esModule: true,
  authSignup: jest.fn(),
}));

describe('Signup', () => {
  const renderComponent = (error = null) => {
    const store = {
      getState: () => {
        const state = { ...defaultState };
        state.ui.error = error;
        return state;
      },
      subscribe: () => {},
      dispatch: () => {},
    };
    return render(
      <Provider store={store}>
        <MemoryRouter>
          <SignupPage />
        </MemoryRouter>
      </Provider>,
    );
  };

  test('snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  test('should dispatch signup action', () => {
    const userData = {
      username: 'user123',
      password: 'supersegurisimo',
      email: 'user123@test.com',
    };

    // Render component
    renderComponent();

    // Test UI components
    const usernameInput = screen.getByLabelText(/Username/);
    const passwordInput = screen.getByLabelText(/Password/);
    const emailInput = screen.getByLabelText(/Email/);
    const submitButton = screen.getByRole('button', { name: /Register/ });
    expect(submitButton).toBeDisabled();

    // Test button after fulfilling text fields
    userEvent.type(usernameInput, userData.username);
    userEvent.type(passwordInput, userData.password);
    userEvent.type(emailInput, userData.email);
    expect(submitButton).toBeEnabled();
    userEvent.click(submitButton);

    expect(authSignup).toHaveBeenCalledWith(userData);
  });

  test('should display an error', () => {
    const resetErrorSpy = jest.spyOn(
      require('../../../../store/slices/ui'),
      'resetError',
    );

    const email = 'example@example.com';
    const error = {
      data: { error: `Email: ${email} is already registered!` },
    };

    renderComponent(error);
    const errorElement = screen.getByText(error.data.error);

    expect(errorElement).toBeInTheDocument();
    const modalButton = screen.getByTestId('modalSignup');
    userEvent.click(modalButton);

    expect(resetErrorSpy).toHaveBeenCalled();
  });
});
