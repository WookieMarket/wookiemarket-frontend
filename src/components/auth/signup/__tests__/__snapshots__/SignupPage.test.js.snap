// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup snapshot 1`] = `
<div>
  <div>
    <header
      class="header"
    >
      <nav
        class="navbar"
      >
        <div
          class="icons"
        >
          <a
            href="/"
          >
            <h1
              class="title"
            >
               
              header
               
            </h1>
          </a>
          <img
            alt="ES"
            class="icon-language"
            src="/images/languageIcons/es.png"
            title="icon-hover"
          />
          <img
            alt="EN"
            class="icon-language"
            src="/images/languageIcons/en.png"
            title="icon-hover"
          />
        </div>
        <ul>
          <li>
            <a
              class="navbar-list-item"
              href="/create-ad"
            >
              Upload ad
            </a>
            <a
              class="navbar-list-item"
              href="/ads"
            >
              Ads
            </a>
            <a
              class="navbar-list-item"
              href="/login"
            >
              Login
            </a>
            <a
              class="navbar-list-item"
              href="/signup"
            >
              Sign up
            </a>
          </li>
        </ul>
      </nav>
      <hr />
    </header>
    <main
      class="content"
    >
      <h2
        class="layout-title"
      >
        Signup Page
      </h2>
      <div>
        <form
          class="container-form"
        >
          <div
            class="email-label"
          >
            <label
              for="email"
            >
              Email
            </label>
            <input
              class="email-input"
              id="email"
              name="email"
              placeholder="Email"
              type="text"
              value=""
            />
          </div>
          <div
            class="user-label"
          >
            <label
              for="username"
            >
              Username
            </label>
            <input
              class="user-input"
              id="username"
              name="username"
              placeholder="Username"
              type="text"
              value=""
            />
          </div>
          <div
            class="password-label"
          >
            <label
              for="password"
            >
              Password
            </label>
            <input
              class="password-input"
              id="password"
              name="password"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <button
            data-testid="button"
            disabled=""
            type="submit"
            width="button-form"
          >
            Register-button-title
          </button>
        </form>
      </div>
    </main>
    <footer>
      <hr />
      <div
        class="footer"
      >
        <p>
          footer
        </p>
      </div>
    </footer>
  </div>
</div>
`;
